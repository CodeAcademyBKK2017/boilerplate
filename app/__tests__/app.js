import apiNotes from '../api';
import App from '../app';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {Alert, AsyncStorage} from 'react-native';
import {shallow} from 'enzyme';

jest.mock('../api'); // file mock

jest.mock('AsyncStorage', () => ({
  getItem: jest.fn(() => Promise.resolve('')),
  setItem: jest.fn(() => Promise.resolve())
}));

jest.mock('uuid', () => () => 'some uuid');

describe('App', () => {

  const props = {};
  const wrapper = shallow(<App {...props}/>);
  const instance = wrapper.instance();
      
  const stateClear = {
    modalData: {},
    titleText: '',
    contentText: '',
    NOTES: []
  };

  it('renders correctly', () => {
    const tree = renderer.create(<App />);
    expect(tree).toBeDefined();
    expect(tree).toMatchSnapshot();
  });

  it('Check state', () => {
    expect(instance.state.titleText).toEqual('');
    expect(instance.state.contentText).toEqual('');
  });

  it('Check Function onContentChange', () => {
    instance.onContentChange('some');
    expect(instance.state.contentText).toEqual('some');
    expect(instance.state.contentText.length).toEqual(4);
  });
  
  it('Check Function onTitleChenge', () => {
    instance.onTitleChange('title');
    expect(instance.state.titleText).toEqual('title');
  });

  it('Check Function showFlatList', () => {
    instance.onTitleChange('React Native');
    instance.onContentChange('- UI');
    instance.onSave();
    expect(instance.showFlatList).toMatchSnapshot();
  }); 

  it('Check Function viewOverlay', () => {
    const note = {
      title: 'React Native',
      content: '- UI',
      key: 0
    };
    instance.onShowModal(note);
    expect(instance.viewOverlay).toMatchSnapshot();
  });

  it('Check Function onShowModal', () => {
    const note = {
      title: 'React Native',
      content: '- UI',
      key: 0
    };
    instance.onShowModal(note)();
    expect(instance.state.modalData).toEqual(note);
  });

  it('Check Function onCloseModal', () => {
    instance.onCloseModal();
    expect(instance.state.modalData).toEqual({});
  });

  it('Check Function openAbout with mock', () => { // mock
    const props = {
      navigation: {
        navigate: jest.fn()
      }
    };
    const wrapper = shallow(<App {...props}/>);
    const instance = wrapper.instance();
    instance.openAbout();
    expect(instance.props.navigation.navigate).toHaveBeenCalledWith('About');
  });

  it('Check Function openAbout with spy', () => { // spy
    const props = {
      navigation: {
        navigate: () => {}
      }
    };
    const appComp = <App {...props}/>;
    const wrapper = shallow(appComp);
    const instance = wrapper.instance();
    const spyFunc = jest.spyOn(props.navigation, 'navigate');

    instance.openAbout();
    
    expect(spyFunc).toHaveBeenCalledWith('About');
  });

  it('componentDidMount with existed notes', () => {
    jest.spyOn(instance, 'onLoadDataState');
    instance.componentDidMount();    
    expect(instance.onLoadDataState).toBeCalled();
  });

  it('onLoadDataState', async () => { // connect server success !!

    instance.setState(stateClear); // clear state
    
    const notes = await apiNotes.getNotes();
    instance.setState({NOTES: notes});

    const expected = {
      modalData: {},
      titleText: '',
      contentText: '',
      NOTES: [{
        title: 'React Native',
        content: '- UI',
        key: 0,
        id: 1
      }]
    };
    expect(instance.state).toEqual(expected);
  });

  xit('Check Function onSave success', async () => {

    instance.setState(stateClear); // clear state

    const title = 'React Native';
    const content = '- UI';
    instance.setState({titleText: title, contentText: content});

    await instance.onSave();

    const expectedNote = {
      'title': 'React Native',
      'content': '- UI',
      'key': 0
    };

    const expected = {
      modalData: {},
      titleText: '',
      contentText: '',
      NOTES: [{
        title: 'React Native',
        content: '- UI',
        key: 0,
        id: 1
      }]
    };

    expect(apiNotes.addNotes).toHaveBeenLastCalledWith(expectedNote);
    expect(AsyncStorage.setItem).toHaveBeenLastCalledWith('state', JSON.stringify(expected));
    expect(instance.state).toEqual(expected);
  });

  it('Check Function onSave failure', async () => {
    await instance.onSave();
    apiNotes.addNotes.mockClear();
    AsyncStorage.setItem.mockClear();
    apiNotes.addNotes.mockImplementation(() => Promise.reject('API failed'));
    expect(AsyncStorage.setItem).not.toBeCalled();
  });

  xit('Check Function onDelete', async () => {

    instance.setState(stateClear); // clear state

    const note = [{
      title: 'React Native',
      content: '- UI',
      key: 0,
      id: 1
    }];
    
    console.log(instance.state);

    instance.setState({NOTES: note});

    console.log(instance.state);

    instance.onDelete(note);

    console.log(instance.state);

    expect(apiNotes.deleteNotes).toBeCalled();
    expect(instance.state).toEqual(stateClear);
  });

  xit('onDeletePress failure', async () => {
    apiNotes.deleteNote.mockClear();
    AsyncStorage.setItem.mockClear();
    apiNotes.deleteNote.mockImplementation(() => Promise.reject('API failed'));
    await instance.onDeletePress()();
    expect(AsyncStorage.setItem).not.toBeCalled();
    expect(Alert.alert).toBeCalled();
  });
});
