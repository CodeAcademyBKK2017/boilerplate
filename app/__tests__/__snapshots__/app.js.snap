// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Check Function showFlatList 1`] = `[Function]`;

exports[`App Check Function viewOverlay 1`] = `[Function]`;

exports[`App renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingTop": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 1,
            "fontWeight": "bold",
            "justifyContent": "flex-start",
          }
        }
      >
        Note Title
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#EEEEEE",
            "borderColor": "gray",
            "borderRadius": 5,
            "borderWidth": 1,
            "padding": 8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          en
        </Text>
      </View>
    </View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Tasks for today"
      placeholderTextColor="#000"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 3,
          "borderWidth": 1,
          "fontWeight": "bold",
          "justifyContent": "flex-start",
          "padding": 10,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "paddingBottom": 15,
          "paddingTop": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "fontStyle": "italic",
            "justifyContent": "flex-start",
            "marginRight": 5,
          }
        }
      >
        Please type your note below.
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#333333",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "fontcustom",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        ï„¨
      </Text>
    </View>
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="- Understand how react-native works.
- Build a native android and iOS app.
- Setup CI for automated builds.
- Conventions to manage the codebase."
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 3,
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "flex-start",
          "padding": 10,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingTop": 15,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Save
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        0
         chacters
      </Text>
    </View>
  </View>
  <Modal
    animationOutType="fadeOut"
    animationType="none"
    hardwareAccelerated={false}
    onClose={[Function]}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(37, 8, 10, 0.78)",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 0,
          "padding": 20,
        }
      }
      testID={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "padding": 20,
            },
            undefined,
          ]
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#333333",
            }
          }
        />
      </View>
    </View>
  </Modal>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "bottom": 0,
        "display": "flex",
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      About Us
    </Text>
  </View>
</View>
`;
